cuqi/model/_model.py

```diff
@@ -336,8 +337,16 @@ def forward(self, *args, is_par=True, **kwargs):
             if x.dim != self.domain_dim:
                 raise ValueError("Attempting to match parameter name of Model with given distribution, but distribution dimension does not match model domain dimension.")
             new_model = copy(self)
-            new_model._non_default_args = [x.name] # Defaults to x if distribution had no name
+            new_model._non_default_args = [x.par_name] # Defaults to x if distribution had no name
             return new_model
+        
+        # If input is a random variable, we handle it separately
+        if isinstance(x, cuqi.randomvariable.RandomVariable):
+            return self._handle_random_variable(x)
+        
+        # If input is a Node from internal abstract syntax tree, we let the Node handle the operation
```

Comment:
Actually this is not the same as raising a NotImplementedError. 

This is a signalling tool to indicate that `Model` does not have the operation implemented, but without raising an error just yet. This is usueful e.g. for stuff like:

```python
x = Gaussian(0,1).rv
y = A@(x+1)
```

Here by convention, Python will first request `A` for its `__matmul__` implementation. If it is implemented it will call it. If not it calls `__rmatmul` of (x+1).

In `model` we only support non-transformed random variables as we can change the name of the input parameter in that case. For more complicated cases, we have to rely on the general implementation of the RV class.

So, for Transformed Random Variables like the (x+1) above, we want to return `NotImplemented` to indicate that `x+1` should handle the matrix multiplication usings its `__rmatmul__` method.